graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,356 @@
+                    try {
+                        if (graphqlSubscriptions.containsKey(id)) {
+                            throw new IllegalArgumentException(
+                                    "Subscription with id " + id + " already exists");
+                        }
+                        final String operationName = toStringFromJson("operationName",
+                                                                      payload.get("operationName"));
+                        final String query = toStringFromJson("query", payload.get("query"));
+                        final Map<String, Object> variables = toMapFromJson(payload.get("variables"));
+                        final Map<String, Object> extensions = toMapFromJson(payload.get("extensions"));
+
+                        final ExecutionInput.Builder executionInput =
+                                ExecutionInput.newExecutionInput()
+                                              .graphQLContext(upgradeCtx)
+                                              .graphQLContext(
+                                                      connectionCtx)
+                                              .query(query)
+                                              .variables(variables)
+                                              .operationName(
+                                                      operationName)
+                                              .extensions(extensions);
+
+                        final CompletableFuture<ExecutionResult> future =
+                                graphqlExecutor.executeGraphql(ctx, executionInput);
+
+                        future.handle(((executionResult, throwable) -> {
+                            handleBlocking(out, id, executionResult, throwable);
+                            return null;
+                        }));
+                    } catch (Exception e) {
+                        logger.debug("Error handling subscription", e);
+                        GraphqlWSSubProtocol.this.writeError(out, id, e);
+                        return;
+                    }
+                    break;
+                case "complete":
+                    ensureInitiated();
+                    // Read id and remove that subscription
+                    id = toStringFromJson("id", eventMap.get("id"));
+                    requireNonNull(id, "id");
+                    final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                    if (s != null) {
+                        s.setCompleted();
+                    }
+                    return;
+                default:
+                    throw new IllegalArgumentException("Unknown event type: " + type);
+            }
+        } catch (Exception e) {
+            logger.debug("Error while handling event", e);
+            out.close(e);
+        }
+    }
+
+    private void handleBlocking(WebSocketWriter out, String id,
```

Comment:
Naming) This isn't handled from a blocking thread which I think can confuse others. What do you think of `handleExecutionResult`?

```suggestion
    private void handleExecutionResult(WebSocketWriter out, String id,
```